File: LeetcodeQuestions/flatten_nested_list_iterator-341.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
# class NestedInteger(object):
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """

class NestedIterator(object):

    def __init__(self, nestedList):
        """
        Initialize your data structure here.
        :type nestedList: List[NestedInteger]
        """
        self.list = nestedList
        self.idx = -1
        self.flatten_list = []
        self.flatten(self.list)

    def next(self):
        """
        :rtype: int
        """
        self.idx += 1
        return self.flatten_list[self.idx]

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.idx < len(self.flatten_list) - 1

    def flatten(self, list):
        for i in range(len(list)):
            current = list[i]
            if current.isInteger():
                self.flatten_list.append(current.getInteger())
            else:
                self.flatten(current.getList())


# Your NestedIterator object will be instantiated and called as such:
# i, v = NestedIterator(nestedList), []
# while i.hasNext(): v.append(i.next())

File: LeetcodeQuestions/flatten_nested_list_iterator-341.rb
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
#
#class NestedInteger
#    def is_integer()
#        """
#        Return true if this NestedInteger holds a single integer, rather than a nested list.
#        @return {Boolean}
#        """
#
#    def get_integer()
#        """
#        Return the single integer that this NestedInteger holds, if it holds a single integer
#        Return nil if this NestedInteger holds a nested list
#        @return {Integer}
#        """
#
#    def get_list()
#        """
#        Return the nested list that this NestedInteger holds, if it holds a nested list
#        Return nil if this NestedInteger holds a single integer
#        @return {NestedInteger[]}
#        """

class NestedIterator
    # @param {NestedInteger[]} nested_list
    def initialize(nested_list)
        @flatten_list = []
        @idx = -1
        @list = nested_list
        self.flatten(@list)
    end

    # @return {Boolean}
    def has_next
        @idx < @flatten_list.length - 1
    end

    # @return {Integer}
    def next
        @idx += 1
        return @flatten_list[@idx]
    end
    
    def flatten(list)
        list.each do |current|
            if current.is_integer()
                @flatten_list << current.get_integer()
            else
                self.flatten(current.get_list())
            end
        end
    end
end

# Your NestedIterator will be called like this:
# i, v = NestedIterator.new(nested_list), []
# while i.has_next()
#    v << i.next
# end